nx/lib/nx/lin_alg.ex

```diff
@@ -967,4 +967,96 @@ defmodule Nx.LinAlg do
       opts
     )
   end
+
+  @doc """
+  Produces the tensor taken to the given power by dot-product.
+
+  The input is always a square tensor and a non-negative integer,
+  and the output is a square tensor of the same dimensions as the input tensor.
+
+  The dot-products are unrolled inside `defn`.
+
+  ## Examples
+
+      iex> Nx.LinAlg.dot_power(Nx.tensor([[1, 2], [3, 4]]), 0)
+      #Nx.Tensor<
+        s64[2][2]
+        [
+          [1, 0],
+          [0, 1]
+        ]
+      >
+
+      iex> Nx.LinAlg.dot_power(Nx.tensor([[1, 2], [3, 4]]), 6)
+      #Nx.Tensor<
+        s64[2][2]
+        [
+          [5743, 8370],
+          [12555, 18298]
+        ]
+      >
+
+      iex> Nx.LinAlg.dot_power(Nx.eye(3), 65535)
+      #Nx.Tensor<
+        s64[3][3]
+        [
+          [1, 0, 0],
+          [0, 1, 0],
+          [0, 0, 1]
+        ]
+      >
+
+      iex> Nx.LinAlg.dot_power(Nx.tensor([[1, 2], [3, 4]]), -1)
+      #Nx.Tensor<
+        f32[2][2]
+        [
+          [-2.0, 1.0],
+          [1.5, -0.5]
+        ]
+      >
+
+      iex> Nx.LinAlg.dot_power(Nx.tensor([[1, 2], [3, 4], [5, 6]]), 1)
+      ** (ArgumentError) expected tensor to match shape {x, x}, got tensor with shape {3, 2}
+  """
+  @doc from_backend: false
+  def dot_power(tensor, power) when is_integer(power) and power < 0 do
+    dot_power(Nx.LinAlg.invert(tensor), -power)
+  end
+
+  def dot_power(tensor, 0) do
+    # We need a special-case for 0 since the code below
+    # is optimized to not compute an initial eye.
+    Nx.eye(tensor)
+  end
+
+  def dot_power(tensor, 1) do
+    # We also need a special case for 1, because the code
+    # below is optimized for calculating powers for
+    # integers where `length(Integer.digits(n, 2)) > 1`.
+    Nx.Defn.Kernel.assert_shape_pattern(tensor, {x, x})
+
+    tensor
+  end
+
+  def dot_power(tensor, power) when is_integer(power) do
+    Nx.Defn.Kernel.assert_shape_pattern(tensor, {x, x})
+
+    {result, exp_tensor} =
+      power
+      |> Integer.digits(2)
+      |> tl()
+      |> Enum.reverse()
+      |> Enum.reduce({nil, tensor}, fn
+        1, {nil, exp_tensor} ->
+          {exp_tensor, Nx.dot(exp_tensor, exp_tensor)}
+
+        1, {result_tensor, exp_tensor} ->
+          {Nx.dot(result_tensor, exp_tensor), Nx.dot(exp_tensor, exp_tensor)}
+
+        0, {result_tensor, exp_tensor} ->
+          {result_tensor, Nx.dot(exp_tensor, exp_tensor)}
+      end)
+
+    Nx.dot(result, exp_tensor)
```

Comment:
```suggestion
      power
      |> Integer.digits(2)
      |> tl()
      |> Enum.reverse()
      |> Enum.reduce({nil, tensor}, fn
        1, {nil, exp_tensor} ->
          {exp_tensor, Nx.dot(exp_tensor, exp_tensor)}

        1, {result_tensor, exp_tensor} ->
          {Nx.dot(result_tensor, exp_tensor), Nx.dot(exp_tensor, exp_tensor)}

        0, {result_tensor, exp_tensor} ->
          {result_tensor, Nx.dot(exp_tensor, exp_tensor)}
      end)
      |> then(fn {result, exp_tensor} -> 
        Nx.dot(result, exp_tensor)
      end)
```

WDYT?

Reply:
The diff ended up with low readability. I basically removed the hard-match in favor of using `Kernel.then` to keep everything in a pipe

Reply:
I came to the realization that I made a mistake with the extra 1-case (specifically, it fails for all `power=2^n`. I'll take your suggestion, and alter it slightly to fix that mistake.